source/opt/fused_multiply_add_pass.h

```diff
@@ -0,0 +1,43 @@
+
+
+#ifndef SOURCE_OPT_FUSED_MULTIPLY_ADD_H_
+#define SOURCE_OPT_FUSED_MULTIPLY_ADD_H_
+
+#include <cstdio>
+#include <memory>
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "source/opt/function.h"
+#include "source/opt/pass.h"
+#include "source/opt/type_manager.h"
+
+namespace spvtools {
+namespace opt {
+
+// Documented in optimizer.hpp
+class FusedMultiplyAddPass : public Pass {
+ public:
+  FusedMultiplyAddPass() {}
+
+  const char* name() const override { return "fused-multiply-add-pass"; }
+
+  Status Process() override;
+
+  IRContext::Analysis GetPreservedAnalyses() override {
+    return IRContext::kAnalysisDefUse |
+           IRContext::kAnalysisInstrToBlockMapping |
```

Comment:
I see the same constants being used in `CopyPropagateArrays` for instance. What do you mean they are not kept up to date?

Reply:
These two analysies are not automatically kept up-to-date.  If the passes changes the code, the pass needs to update to reflect the change.  You can look at https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/ir_builder.h#L606  to see how the updates can happen.

Your code creates a new instruction.  It is not analyzed by the def-use-manager, or added to the inst-to-block mapping.  This would need to happen at line 63 of fused_multiply_add_pass.cpp.  Alternatively you could use the instruction builder.  Here is an example on how to use it: https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/amd_ext_to_khr.cpp#L71

Reply:
I have a version now that uses `AddNaryExtendedInstruction`. Since this function only takes `vector<uint32_t>` instead of `Instruction::OperandList`, I need to copy over the operands from the old instruction to the new instruction one by one. Moreover, the `AddOperand` function only accepts move semantics, which bloats the code a bit. Is there a particular reason why there is no overloaded version of this function that accepts a constant reference?

And thank you for the feedback so far, I appreciate the assistance as I'm still taking over the work on our changes in various third party libraries.

Reply:
No there is no particular reason.  Probably just that it was not needed.  You can add the const ref version of you would like.